Fix: Improves performance (refs #3530)

mysticatea opened the pull request:
This was separated from #3559 

- `RuleContext` method definitions are moved to outside of the constructor. `RuleContext` object is created many times, so that constructor should be as simple as possible. (`lib/rule-context.js`)
- Construction of a regular expression object which depends on rule's options is moved to outside of `getNodeIndent` function. The function is called many times and to create dynamic regular expression object is slow. (`lib/rules/indent.js`)
- Avoid calling `Function#apply` as possible. (`lib/eslint.js` and `lib/rule-context.js`)
- Avoid calling `SourceCode#getTokens`. A CPU Profiler said that function is slow a bit. (`lib/rules/no-spaced-func.js` and `lib/rules/space-before-function-paren.js`)

**Before** (`npm run perf` for multiple files)

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  3882.190459ms
Performance Run #2:  3767.144016ms
Performance Run #3:  3813.567308ms
Performance Run #4:  3784.897383ms
Performance Run #5:  3768.8099039999997ms
Performance budget exceeded: 3784.897383ms (limit: 2211.0849056603774ms)


Load performance Run #1:  252.483815ms
Load performance Run #2:  251.585219ms
Load performance Run #3:  248.925052ms
Load performance Run #4:  252.786566ms
Load performance Run #5:  253.870494ms

Load Performance median :  252.483815ms
```

**After**

```
CPU Speed is 3392 with multiplier 7500000
Performance Run #1:  3075.235085ms
Performance Run #2:  3030.254616ms
Performance Run #3:  3057.690596ms
Performance Run #4:  3047.723645ms
Performance Run #5:  3016.89371ms
Performance budget exceeded: 3047.723645ms (limit: 2211.0849056603774ms)


Load performance Run #1:  254.633259ms
Load performance Run #2:  250.226913ms
Load performance Run #3:  253.68184ms
Load performance Run #4:  252.021688ms
Load performance Run #5:  251.831827ms

Load Performance median :  252.021688ms
```

lo1tuma commented on lib/rule-context.js:
```diff
@@ -72,42 +72,27 @@ var PASSTHROUGHS = [
+
+    // private.
+    this.eslint = eslint;
+    this.severity = severity;
```
I'm a bit concerned about these changes. This would allow rules to modify their own id or the severity.

mysticatea replied:
I tried to add `Object.freeze(this)` into the end of this constructor, then `npm run perf` (target is `lib`) come to slow about 60ms (2%).

Hmm, I will try to detect that with `RuleTester`.

mysticatea replied:
Oh...
I tried to freeze `context`, `context.options`, `context.settings`, and `context.ecmaFeatures` in `RuleTester`.
Then, `npm test` throws 149 errors...

nzakas replied:
Agree with @lo1tuma , we need to keep his these work the same. In my tests, you can speed things up by defining the property first using `this.foo =` and then using `Object.defineProperty`.

mysticatea replied:
I tried:

|                             | Time   |
|-----------------------------|-------:|
| Assignments                 | 3010ms |
| Assignments + Object.freeze | 3070ms |
| Object.defineProperties     | 3362ms |
| Object.defineProperty x 6   | 3254ms |

I made `RuleTester` freezing those, but is it not enough?
If so, I'm going to use `Object.freeze`.

nzakas replied:
I don't think `RuleTester` is enough, so let's go with `Object.freeze()`.

mysticatea commented:
I updated with more optimizations.

- To search variables by name exists in many places. And currently the search is linear search. Especially, browser environment has many many variables. So I replaced those with O(1) search.
- `RuleTester` comes to check whether or not `RuleContext` instance was not modified by rules.

As a result, in my PC:

|                                          | before | after  | %    |
|------------------------------------------|-------:|-------:|-----:|
| `npm test` time                          | 52s    | 34s    | -35% |
| `npm run perf` (single)                  | 2204ms | 2182ms |  -1% |
| `npm run perf` (multi)                   | 3804ms | 3013ms | -21% |
| `npm run perf` (single with browser env) | 2266ms | 2182ms |  -4% |
| `npm run perf` (multi with browser env)  | 5450ms | 3634ms | -33% |

- single: `--no-ignore ./tests/performance/jshint.js`
- multi: `--no-ignore lib`
- single with browser env: `--no-ignore ./tests/performance/jshint.js --env browser`
- multi with browser env: `--no-ignore lib --env browser`

nzakas commented on lib/eslint.js:
```diff
@@ -891,8 +874,20 @@ module.exports = (function() {
+            }
+
+            // Avoid `Function.prototype.apply` because of slow.
+            switch (arguments.length) {
```
You might be able to speed this up by avoiding `arguments`

nzakas replied:
Yeah, so, try this:

```js
api[methodName] = function(a, b, c, d, e) {
   // ...

   return sourceCode[exMethodName](a, b, c , d, e)
};
```

I don't think we have any methods with more than five arguments.

mysticatea replied:
Oh sorry, yesterday I could not get enough time yesterday. I will try it today!

mysticatea replied:
Result: `3030ms` → `3035ms`.

Hmm, it's wonder.
`return sourceCode[exMethodName](a, b, c , d, e)` is simple and understandable, so I will use this.

mysticatea commented:
I updated with `Object.freeze(this)` and `(a, b, c, d, e)`!

nzakas commented on lib/rule-context.js:
```diff
@@ -72,42 +72,29 @@ var PASSTHROUGHS = [
+    this.id = ruleId;
+    this.options = options;
+    this.settings = settings;
+    this.ecmaFeatures = ecmaFeatures;
```
This was using `Object.create ` before and was frozen.

mysticatea replied:
Hmm, `ecmaFeatures` still can be modified, and 3-4% slower. Is it needed?
If it's needed, should I treat `options` and `settings` with the same way?

※ `RuleTester` is freezing `ecmaFeatures`, `settings`, and `options` deeply.

```js
context.ecmaFeatures.__proto__.modules = true;
```

| `this.ecmaFeatures = ???`                    | Time   |
|----------------------------------------------|-------:|
| `ecmaFeatures`                               | 3004ms |
| `Object.freeze(Object.create(ecmaFeatures))` | 3112ms |
| `assign({}, ecmaFeatures)`                   | 3095ms |
| `Object.freeze(assign({}, ecmaFeatures))`    | 3115ms |

If it treats `settings` and `options` as well, then 9-10% slower. 
`RuleContext` constructor is called many times (rules ✖ files), so this has big impact to performance.

nzakas replied:
Hmmm OK, let's keep it as-is. I wish V8 was faster!
